File: ios/StillMoment/Application/ViewModels/TrimEditorViewModel.swift
//
//  TrimEditorViewModel.swift
//  Still Moment
//
//  Application Layer - Waveform Trim Editor ViewModel (shared-107)
//

import Combine
import Foundation
import OSLog

/// How long the short auto-previews play after committing a mark.
struct TrimPreviewDurations {
    let afterMarkDrag: TimeInterval
    let afterNudge: TimeInterval

    static let standard = TrimPreviewDurations(afterMarkDrag: 2.2, afterNudge: 1.4)
}

/// ViewModel for the full-screen waveform trim editor.
///
/// Owns the immutable `TrimEditorState` and forwards intents (select, move, nudge,
/// whole-file) to it. Loads the waveform lazily through `WaveformProviderProtocol`
/// and drives audio playback through the shared meditation-preview infrastructure
/// (shared-098, `.preview` session).
///
/// Playback model (handoff "touch-robuste Punkt-Bedienung"): the playhead is its own,
/// always-present position. Dragging it (`seek`) pauses playback first and moves only
/// the playhead. Releasing a mark drag or nudging anchors the playhead at the mark and
/// auditions the cut with a short auto-preview; ▶ plays from the playhead, ⏸ pauses and
/// keeps it. Playback pauses automatically at the end point, unless it was anchored at
/// or after the end point (so the end position itself can be auditioned).
@MainActor
final class TrimEditorViewModel: ObservableObject {
    // MARK: Lifecycle

    init(
        meditation: GuidedMeditation,
        audioService: AudioServiceProtocol = AudioService(),
        waveformProvider: WaveformProviderProtocol = WaveformProvider(),
        meditationService: GuidedMeditationServiceProtocol = GuidedMeditationService(),
        previewDurations: TrimPreviewDurations = .standard
    ) {
        self.meditation = meditation
        self.audioService = audioService
        self.waveformProvider = waveformProvider
        self.meditationService = meditationService
        self.previewDurations = previewDurations

        let state = TrimEditorState(meditation: meditation)
        self.editorState = state
        self.playheadTime = state.start

        self.bindPreviewPublishers()
    }

    // MARK: Internal

    @Published private(set) var editorState: TrimEditorState
    @Published private(set) var waveform: MeditationWaveform?
    @Published private(set) var waveformLoadFailed = false

    /// Whether audio is currently playing via ▶ — drives the ▶/⏸ icon.
    @Published private(set) var isPlaying = false
    /// Whether a short auto-preview (after mark drag/nudge) is running. The play
    /// button stays calm (▶) during previews.
    @Published private(set) var isPreviewing = false
    /// Playback position in seconds — always present, seeded with the start point.
    @Published private(set) var playheadTime: TimeInterval

    /// True while the waveform is still being generated/loaded and has not failed.
    var isLoadingWaveform: Bool {
        self.waveform == nil && !self.waveformLoadFailed
    }

    // MARK: - Waveform Loading

    /// Loads the waveform through the provider (cache hit is instant, miss generates).
    /// Runs in a stored task so the view can release it on disappear; the provider's
    /// shared generation keeps running and caches even if we stop awaiting (intended).
    func loadWaveform() {
        guard self.waveform == nil, !self.waveformLoadFailed else {
            return
        }
        self.waveformLoadTask?.cancel()
        self.waveformLoadTask = Task { [weak self] in
            guard let self else {
                return
            }
            do {
                let loaded = try await self.waveformProvider.waveform(for: self.meditation)
                guard !Task.isCancelled else {
                    return
                }
                self.waveform = loaded
            } catch is CancellationError {
                // Editor closed mid-load — nothing to show, no fallback needed.
            } catch {
                Logger.viewModel.error("Failed to load waveform for trim editor", error: error)
                self.waveformLoadFailed = true
            }
        }
    }

    // MARK: - Editor Intents

    /// Selects which point (start/end) is active and moves the playhead onto it.
    /// Keeps playing (seek) when playing, stays paused otherwise.
    func selectPoint(_ point: TrimPoint) {
        self.cancelPreview()
        self.editorState = self.editorState.selecting(point)
        self.anchorPlayhead(at: self.editorState.activeValue)
    }

    /// Moves a point to a time (clamped + min-distance enforced by the domain) and selects it.
    /// Fires continuously during a drag — audio only follows on `markDragEnded()`.
    func movePoint(_ point: TrimPoint, to time: TimeInterval) {
        self.cancelPreview()
        self.editorState = self.editorState.moving(point, to: time)
    }

    /// Mark drag released — anchors the playhead at the mark and auditions the cut.
    func markDragEnded() {
        self.playPreview(from: self.editorState.activeValue, for: self.previewDurations.afterMarkDrag)
    }

    /// Nudges the active point by a delta (±1 s) and auditions the new cut.
    func nudgeActivePoint(by delta: TimeInterval) {
        self.editorState = self.editorState.nudgingActivePoint(by: delta)
        self.playPreview(from: self.editorState.activeValue, for: self.previewDurations.afterNudge)
    }

    /// Resets the selection to the full file and parks the playhead at 0, paused.
    func useWholeFile() {
        self.cancelPreview()
        self.pausePlayback()
        self.editorState = self.editorState.usingWholeFile()
        self.playheadTime = 0
        self.playsToFileEnd = false
    }

    // MARK: - Seeking (playhead lane / upper zone)

    /// Moves the playhead while the finger drags in the playhead zone. Marks stay
    /// untouched; a running playback or preview is paused first so it cannot fight
    /// the finger (handoff: every seek pauses first).
    func seek(to time: TimeInterval) {
        self.cancelPreview()
        self.pausePlayback()
        self.playheadTime = min(max(time, 0), self.editorState.duration)
        self.playsToFileEnd = self.playheadTime >= self.editorState.end
    }

    // MARK: - Playback

    /// ▶ plays from the playhead, ⏸ pauses and keeps the position. Pressing ▶ during
    /// a short preview promotes it to full playback from the preview's anchor.
    func togglePlayback() {
        if self.isPlaying {
            self.pausePlayback()
            return
        }
        self.cancelPreview()
        self.startPlayback()
    }

    /// Stops all audio — called when the editor disappears.
    func viewDisappeared() {
        self.previewTask?.cancel()
        self.previewTask = nil
        self.audioService.stopMeditationPreview()
        self.isPlaying = false
        self.isPreviewing = false
        self.waveformLoadTask?.cancel()
        self.waveformLoadTask = nil
    }

    // MARK: Private

    private let meditation: GuidedMeditation
    private let audioService: AudioServiceProtocol
    private let waveformProvider: WaveformProviderProtocol
    private let meditationService: GuidedMeditationServiceProtocol
    private let previewDurations: TrimPreviewDurations

    private var waveformLoadTask: Task<Void, Never>?
    private var previewTask: Task<Void, Never>?
    private var cancellables = Set<AnyCancellable>()
    /// True when the current playback was anchored at/after the end point — it then runs
    /// to the file end instead of pausing at the end point (auditioning the end position).
    private var playsToFileEnd = false

    private func bindPreviewPublishers() {
        self.audioService.meditationPreviewPositionPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] position in
                guard let self, self.isPlaying || self.isPreviewing else {
                    return
                }
                self.handlePlaybackPosition(position)
            }
            .store(in: &self.cancellables)

        self.audioService.meditationPreviewCompletionPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] in
                guard let self else {
                    return
                }
                self.previewTask?.cancel()
                self.previewTask = nil
                self.isPlaying = false
                self.isPreviewing = false
            }
            .store(in: &self.cancellables)
    }

    private func handlePlaybackPosition(_ position: TimeInterval) {
        if !self.playsToFileEnd, position >= self.editorState.end {
            self.pauseAtEndPoint()
            return
        }
        self.playheadTime = position
    }

    private func startPlayback() {
        let position = self.playheadTime
        guard self.startPreviewPlayback(at: position) else {
            return
        }
        self.isPlaying = true
        self.playsToFileEnd = position >= self.editorState.end
    }

    /// Pause keeps `playheadTime` so ▶ resumes where the user stopped.
    private func pausePlayback() {
        guard self.isPlaying else {
            return
        }
        self.audioService.stopMeditationPreview()
        self.isPlaying = false
    }

    private func pauseAtEndPoint() {
        self.previewTask?.cancel()
        self.previewTask = nil
        self.audioService.stopMeditationPreview()
        self.isPlaying = false
        self.isPreviewing = false
        self.playheadTime = self.editorState.end
    }

    /// Moves the playhead onto a mark. Keeps playing (seek) when playing,
    /// stays paused (playhead jumps only) otherwise.
    private func anchorPlayhead(at position: TimeInterval) {
        self.playheadTime = position
        self.playsToFileEnd = position >= self.editorState.end
        if self.isPlaying {
            self.audioService.seekMeditationPreview(to: position)
        }
    }

    /// Auditions the cut: plays a short preview from `position`, then stops and parks
    /// the playhead back at `position`. Replaces a running playback or preview.
    private func playPreview(from position: TimeInterval, for duration: TimeInterval) {
        self.cancelPreview()
        self.pausePlayback()
        self.playheadTime = position
        self.playsToFileEnd = position >= self.editorState.end
        guard self.startPreviewPlayback(at: position) else {
            return
        }
        self.isPreviewing = true
        self.previewTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(duration * 1_000_000_000))
            guard !Task.isCancelled else {
                return
            }
            self?.finishPreview(parkingPlayheadAt: position)
        }
    }

    private func finishPreview(parkingPlayheadAt position: TimeInterval) {
        self.audioService.stopMeditationPreview()
        self.isPreviewing = false
        self.playheadTime = position
        self.previewTask = nil
    }

    private func cancelPreview() {
        self.previewTask?.cancel()
        self.previewTask = nil
        guard self.isPreviewing else {
            return
        }
        self.audioService.stopMeditationPreview()
        self.isPreviewing = false
    }

    /// Starts the underlying preview player at the given position. Returns false when the
    /// file could not be resolved (no playback, no state change).
    private func startPreviewPlayback(at position: TimeInterval) -> Bool {
        guard let fileURL = self.meditationService.fileURL(for: self.meditation) else {
            Logger.viewModel.warning(
                "Cannot play — file not found",
                metadata: ["id": self.meditation.id.uuidString]
            )
            return false
        }
        do {
            try self.audioService.playMeditationPreview(fileURL: fileURL)
            self.audioService.seekMeditationPreview(to: position)
            return true
        } catch {
            Logger.viewModel.error("Failed to start trim playback", error: error)
            return false
        }
    }
}

File: android/app/src/main/kotlin/com/stillmoment/presentation/viewmodel/TrimEditorViewModel.kt
package com.stillmoment.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.stillmoment.domain.models.GuidedMeditation
import com.stillmoment.domain.models.MeditationWaveform
import com.stillmoment.domain.models.TrimEditorState
import com.stillmoment.domain.models.TrimPoint
import com.stillmoment.domain.models.TrimZoomWindow
import com.stillmoment.domain.services.AudioServiceProtocol
import com.stillmoment.domain.services.LoggerProtocol
import com.stillmoment.domain.services.WaveformGenerationException
import com.stillmoment.domain.services.WaveformProviderProtocol
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * How long the short auto-previews play after committing a mark (shared-107).
 * Injectable so tests can shorten them. Values match iOS (2.2 s / 1.4 s) in ms.
 */
data class TrimPreviewDurations(
    val afterMarkDragMs: Long,
    val afterNudgeMs: Long
) {
    companion object {
        val Standard = TrimPreviewDurations(afterMarkDragMs = 2_200L, afterNudgeMs = 1_400L)
    }
}

/**
 * UI state for the full-screen waveform trim editor (shared-107/108).
 *
 * Playhead, playing, previewing and the zoom window are UI concerns; the immutable
 * selection lives in [editorState].
 */
data class TrimEditorUiState(
    val editorState: TrimEditorState,
    val waveform: MeditationWaveform? = null,
    val waveformLoadFailed: Boolean = false,
    val isPlaying: Boolean = false,
    val isPreviewing: Boolean = false,
    val playheadTimeMs: Long = 0L,
    val window: ClosedRange<Long> = 0L..0L
) {
    /** True while the window shows less than the whole file (drives minimap/zoom-out). */
    val isZoomed: Boolean
        get() = (window.endInclusive - window.start) < (editorState.duration - BOUNDARY_TOLERANCE_MS)

    /** True while the waveform is still loading and has not failed. */
    val isLoadingWaveform: Boolean
        get() = waveform == null && !waveformLoadFailed

    private companion object {
        const val BOUNDARY_TOLERANCE_MS = 1_000L
    }
}

/**
 * ViewModel for the full-screen waveform trim editor (shared-107/108).
 *
 * Owns the immutable [TrimEditorState] and forwards intents to it. Loads the waveform
 * lazily through [WaveformProviderProtocol] and drives audio through the shared
 * meditation-preview path ([AudioServiceProtocol]).
 *
 * Playback model (mirrors iOS): the playhead is its own, always-present position.
 * Dragging it ([seek]) pauses playback first and moves only the playhead;
 * [playheadDragEnded] starts playback from there. Releasing a mark drag or nudging
 * auditions the cut with a short auto-preview (start: first seconds of the range; end:
 * last seconds up to the mark) and parks the playhead at the mark. Playback pauses at the
 * end point, unless it starts at/after the end point (the escape hatch to listen past the
 * cut). Lock-screen preview behavior is device-only; not unit-tested.
 */
@HiltViewModel
class TrimEditorViewModel
@Inject
constructor(
    private val audioService: AudioServiceProtocol,
    private val waveformProvider: WaveformProviderProtocol,
    private val logger: LoggerProtocol,
    private val previewDurations: TrimPreviewDurations = TrimPreviewDurations.Standard
) : ViewModel() {
    private lateinit var meditation: GuidedMeditation

    private val _uiState = MutableStateFlow(
        TrimEditorUiState(editorState = TrimEditorState(0L, 0L, 0L, TrimPoint.START))
    )
    val uiState: StateFlow<TrimEditorUiState> = _uiState.asStateFlow()

    private var waveformJob: Job? = null
    private var previewJob: Job? = null

    /**
     * True when the current playback was anchored at/after the end point — it then runs
     * to the file end instead of pausing at the end point (auditioning the end position).
     */
    private var playsToFileEnd = false

    init {
        observePreviewPosition()
        observePreviewCompletion()
    }

    /** Seeds the editor from a meditation. Call once when the screen appears. */
    fun loadMeditation(meditation: GuidedMeditation) {
        this.meditation = meditation
        val state = TrimEditorState.fromMeditation(meditation)
        _uiState.value = TrimEditorUiState(
            editorState = state,
            playheadTimeMs = state.start,
            window = 0L..maxOf(state.duration, 0L)
        )
    }

    // MARK: - Waveform Loading

    /** Loads the waveform through the provider (cache hit is instant, miss generates). */
    fun loadWaveform() {
        val current = _uiState.value
        if (current.waveform != null || current.waveformLoadFailed) {
            return
        }
        waveformJob?.cancel()
        waveformJob = viewModelScope.launch {
            try {
                val loaded = waveformProvider.waveform(meditation)
                _uiState.update { it.copy(waveform = loaded) }
            } catch (e: WaveformGenerationException) {
                logger.e(TAG, "Failed to load waveform for trim editor", e)
                _uiState.update { it.copy(waveformLoadFailed = true) }
            }
        }
    }

    // MARK: - Editor Intents

    /** Selects which point is active and moves the playhead onto it. */
    fun selectPoint(point: TrimPoint) {
        cancelPreview()
        val state = _uiState.value.editorState.selecting(point)
        _uiState.update { it.copy(editorState = state) }
        anchorPlayhead(state.activeValue)
    }

    /** Moves a point (clamped + min-distance by the domain) and selects it. */
    fun movePoint(point: TrimPoint, toMs: Long) {
        cancelPreview()
        _uiState.update { it.copy(editorState = it.editorState.moving(point, toMs)) }
    }

    /** Mark drag released — auditions the cut, parks the playhead, recenters the zoom. */
    fun markDragEnded() {
        auditionActivePoint(previewDurations.afterMarkDragMs)
        recenterWindowOnActivePoint()
    }

    /** Nudges the active point by ±1 s, auditions the new cut, keeps the window centered. */
    fun nudgeActivePoint(deltaMs: Long) {
        _uiState.update { it.copy(editorState = it.editorState.nudgingActivePoint(deltaMs)) }
        auditionActivePoint(previewDurations.afterNudgeMs)
        recenterWindowOnActivePoint()
    }

    /** Resets the selection to the full file and parks the playhead at 0, paused. */
    fun useWholeFile() {
        cancelPreview()
        pausePlayback()
        val state = _uiState.value.editorState.usingWholeFile()
        playsToFileEnd = false
        _uiState.update {
            it.copy(
                editorState = state,
                playheadTimeMs = 0L,
                window = wholeFileWindow(state)
            )
        }
    }

    // MARK: - Zoom (shared-108)

    /** Selects the point and frames it in a zoom window. Short files never zoom. */
    fun focusPoint(point: TrimPoint) {
        selectPoint(point)
        val state = _uiState.value.editorState
        _uiState.update {
            it.copy(window = TrimZoomWindow.frame(state.activeValue, point, state.duration))
        }
    }

    /** Zooms back to the overview; marks and playhead untouched. */
    fun zoomOut() {
        _uiState.update { it.copy(window = wholeFileWindow(it.editorState)) }
    }

    /** Moves the zoom window so it is centered on [centerMs]. */
    fun panWindow(centerMs: Long) {
        _uiState.update {
            it.copy(window = TrimZoomWindow.pan(centerMs, it.editorState.duration))
        }
    }

    // MARK: - Seeking (playhead lane)

    /** Moves the playhead while dragging; pauses a running playback/preview first. */
    fun seek(toMs: Long) {
        cancelPreview()
        pausePlayback()
        val state = _uiState.value.editorState
        val clamped = toMs.coerceIn(0L, state.duration)
        playsToFileEnd = clamped >= state.end
        _uiState.update { it.copy(playheadTimeMs = clamped) }
    }

    /** Playhead drag released — playback starts from the new position. */
    fun playheadDragEnded() {
        cancelPreview()
        startPlayback()
    }

    // MARK: - Playback

    /** ▶ plays from the playhead, ⏸ pauses and keeps the position. */
    fun togglePlayback() {
        if (_uiState.value.isPlaying) {
            pausePlayback()
            return
        }
        cancelPreview()
        startPlayback()
    }

    /** Stops all audio — called when the editor disappears. */
    fun viewDisappeared() {
        previewJob?.cancel()
        previewJob = null
        waveformJob?.cancel()
        waveformJob = null
        audioService.stopMeditationPreview()
        _uiState.update { it.copy(isPlaying = false, isPreviewing = false) }
    }

    // MARK: - Private

    private fun wholeFileWindow(state: TrimEditorState): ClosedRange<Long> = 0L..maxOf(state.duration, 0L)

    private fun recenterWindowOnActivePoint() {
        if (!_uiState.value.isZoomed) {
            return
        }
        val state = _uiState.value.editorState
        _uiState.update {
            it.copy(window = TrimZoomWindow.frame(state.activeValue, state.activePoint, state.duration))
        }
    }

    private fun observePreviewPosition() {
        viewModelScope.launch {
            audioService.meditationPreviewPositionFlow.collect { position ->
                val state = _uiState.value
                if (state.isPlaying || state.isPreviewing) {
                    handlePlaybackPosition(position)
                }
            }
        }
    }

    private fun observePreviewCompletion() {
        viewModelScope.launch {
            audioService.meditationPreviewCompletionFlow.collect {
                previewJob?.cancel()
                previewJob = null
                _uiState.update { it.copy(isPlaying = false, isPreviewing = false) }
            }
        }
    }

    private fun handlePlaybackPosition(position: Long) {
        val state = _uiState.value
        if (!playsToFileEnd && position >= state.editorState.end) {
            pauseAtEndPoint()
            return
        }
        _uiState.update { it.copy(playheadTimeMs = position) }
    }

    private fun startPlayback() {
        val position = _uiState.value.playheadTimeMs
        if (!startPreviewPlayback(position)) {
            return
        }
        playsToFileEnd = position >= _uiState.value.editorState.end
        _uiState.update { it.copy(isPlaying = true) }
    }

    private fun pausePlayback() {
        if (!_uiState.value.isPlaying) {
            return
        }
        audioService.stopMeditationPreview()
        _uiState.update { it.copy(isPlaying = false) }
    }

    private fun pauseAtEndPoint() {
        previewJob?.cancel()
        previewJob = null
        audioService.stopMeditationPreview()
        _uiState.update {
            it.copy(isPlaying = false, isPreviewing = false, playheadTimeMs = it.editorState.end)
        }
    }

    /** Moves the playhead onto a mark; keeps playing (seek) when playing. */
    private fun anchorPlayhead(position: Long) {
        playsToFileEnd = position >= _uiState.value.editorState.end
        _uiState.update { it.copy(playheadTimeMs = position) }
        if (_uiState.value.isPlaying) {
            audioService.seekMeditationPreview(position)
        }
    }

    /**
     * Auditions the cut inside the audible window: start plays the first seconds of the
     * range; end plays the last seconds UP TO the mark (never the cut-off audio).
     */
    private fun auditionActivePoint(durationMs: Long) {
        val state = _uiState.value.editorState
        when (state.activePoint) {
            TrimPoint.START -> playPreview(state.start, durationMs, parkAtMs = state.start)
            TrimPoint.END -> {
                val from = maxOf(state.end - durationMs, state.start)
                playPreview(from, durationMs, parkAtMs = state.end)
            }
        }
    }

    private fun playPreview(fromMs: Long, durationMs: Long, parkAtMs: Long) {
        cancelPreview()
        pausePlayback()
        playsToFileEnd = fromMs >= _uiState.value.editorState.end
        _uiState.update { it.copy(playheadTimeMs = fromMs) }
        if (!startPreviewPlayback(fromMs)) {
            return
        }
        _uiState.update { it.copy(isPreviewing = true) }
        previewJob = viewModelScope.launch {
            delay(durationMs)
            finishPreview(parkAtMs)
        }
    }

    private fun finishPreview(parkAtMs: Long) {
        audioService.stopMeditationPreview()
        previewJob = null
        _uiState.update { it.copy(isPreviewing = false, playheadTimeMs = parkAtMs) }
    }

    private fun cancelPreview() {
        previewJob?.cancel()
        previewJob = null
        if (!_uiState.value.isPreviewing) {
            return
        }
        audioService.stopMeditationPreview()
        _uiState.update { it.copy(isPreviewing = false) }
    }

    /** Starts the preview player at [position]. Returns false if no file is set. */
    private fun startPreviewPlayback(position: Long): Boolean {
        audioService.playMeditationPreview(meditation.fileUri)
        audioService.seekMeditationPreview(position)
        return true
    }

    private companion object {
        const val TAG = "TrimEditor"
    }
}
